src/EventListener/PrettyErrorScreenListener.php

```diff
@@ -0,0 +1,312 @@
+        'Contao\\CoreBundle\\Exception\\NoLayoutSpecifiedException'      => 'no_layout_specified',
+        'Contao\\CoreBundle\\Exception\\NoRootPageFoundException'        => 'no_root_page_found',
+        'Contao\\CoreBundle\\Exception\\PageNotFoundException'           => 'page_not_found',
+        'Contao\\CoreBundle\\Exception\\ServiceUnavailableException'     => 'service_unavailable',
+    ];
+
+    /**
+     * Constructor.
+     *
+     * @param bool          $prettyErrorScreens True to render the error screens
+     * @param TwigEngine    $twig               The twig rendering engine
+     * @param ConfigAdapter $config             The config adapter
+     */
+    public function __construct($prettyErrorScreens, TwigEngine $twig, ConfigAdapter $config)
+    {
+        $this->prettyErrorScreens = $prettyErrorScreens;
+        $this->twig               = $twig;
+        $this->config             = $config;
+    }
+
+    /**
+     * Map the exception to an error screen.
+     *
+     * @param GetResponseForExceptionEvent $event The event object
+     *
+     * @todo Log the exception?
+     */
+    public function onKernelException(GetResponseForExceptionEvent $event)
+    {
+        $exception = $event->getException();
+
+        switch (true) {
+            case $exception instanceof AccessDeniedHttpException:
+                $this->renderErrorScreen(403, $event);
+                break;
+
+            case $exception instanceof BadRequestHttpException:
+            case $exception instanceof InternalServerErrorHttpException:
+                $this->checkExceptionChain($event);
+                break;
+
+            case $exception instanceof NotFoundHttpException:
+                $this->renderErrorScreen(404, $event);
+                break;
+
+            case $exception instanceof ServiceUnavailableHttpException:
+                $this->renderMaintenanceScreen($event);
+                break;
+        }
+    }
+
+    /**
+     * Renders the error screen.
+     *
+     * @param GetResponseForExceptionEvent $event The event object
+     */
+    private function renderErrorScreen($type, GetResponseForExceptionEvent $event)
+    {
+        static $processing;
+
+        if (true === $processing) {
+            return;
+        }
+
+        $processing = true;
+
+        if (null !== ($response = $this->renderPageHandler($type, $event))) {
+            $event->setResponse($response);
+        } else {
+            $this->checkExceptionChain($event);
+        }
+
+        $processing  = false;
+    }
+
+    /**
+     * Renders a Contao page handler.
+     *
+     * @param string $type The error type
+     *
+     * @return Response|null The response object or null
+     */
+    private function renderPageHandler($type)
+    {
+        $type = "error_$type";
+
+        if (!isset($GLOBALS['TL_PTY'][$type]) || !class_exists($GLOBALS['TL_PTY'][$type])) {
```

Comment:
Correct, but rendering them like this in the legacy way is still ugly.
I still would prefer providing route names and sub requests.